### PR TITLE
mountlib: check if directory is not empty before mounting - fixes #1386

### LINKED ISSUE #1386: rclone mount worked in a directory with existing files, causing problems on macOS
If you've just got a question or aren't sure if you've found a bug then please use the [rclone forum](https://forum.rclone.org/) instead of filing an issue.

> What is your rclone version (eg output from `rclone -V`)

v1.36

> Which OS you are using and how many bits (eg Windows 7, 64 bit)

macOS Sierra Version 10.12.4
FUSE for macOS v3.5.6

> Which cloud storage system are you using? (eg Google Drive)

Amazon Cloud Drive, but any should work in this case

> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone mount -vv ACD_crypt:/ ~/.`

> A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

I lost the output but everything worked when it should have failed because the directory was not empty.

I mounted my cloud drive to my home directory by mistake, I intended to mount it to `~/Mounts/acd`.  The mount worked, however, despite the directory not being empty.  This caused Finder to panic and reset all of my preferences for the application dock, Finder file browser, icons, wallpaper, etc.  It was like I just turned on a new computer or created a new user account.  I received an error message from MySQL Workbench stating that it could not write to a file in my home directory.  I then unmounted the rclone drive, restarted the computer, but it was still in the corrupted state.  I had to restore my user directory from a Time Machine backup to get everything back to a good state.  Existing files were still present such as Documents and Downloads, but a condition was created where normal access to my home directory was no longer available and other applications that depend on it panicked and reset to defaults.

This item has already been discussed on the forum at: https://forum.rclone.org/t/rclone-mount-clobbered-my-home-directory-after-a-typing-mistake/2032

### PR DESCRIPTION
**Pull request by xerrni**
Signed-off-by: Ernest Borowski <er.borowski@gmail.com>

### CODE REVIEW

File: cmd/mount/mount.go [hunk 1]
```diff
@@ -73,6 +74,29 @@ func mountOptions(device string) (options []fuse.MountOption) {
 // report an error when fusermount is called.
 func mount(f fs.Fs, mountpoint string) (*vfs.VFS, <-chan error, func() error, error) {
 	fs.Debugf(f, "Mounting on %q", mountpoint)
+
```
Comment (ncw) [inline cmd/mount/mount.go]: Can you factor this bit of code into a function and place it in cmd/mountlib/mount.go - say checkMountEmpty() or something like that.

Then you can call it [about here](https://github.com/ncw/rclone/blob/ed91d6b5a5ffd71c58572d0723afb24554aa2935/cmd/mountlib/mount.go#L140) and the check will work for both mount and cmount (cmount uses libfuse but is called mount on windows)

File: cmd/mount/mount.go [hunk 2]
```diff
@@ -73,6 +74,29 @@ func mountOptions(device string) (options []fuse.MountOption) {
...
+	// Check is folder is empty
+	if !mountlib.AllowNonEmpty {
+		fp, fpErr := os.Open(mountpoint)
+		if fpErr != nil {
+			fp.Close()
```
Comment (ncw) [inline cmd/mount/mount.go]: The build is complaining about you not checking the error from close.

If you factor this into a function then you can write

    defer fs.CheckClose(fp, &err)

after the open (you'll have to name the error return parameter).

File: cmd/mount/mount.go (hunk 1 above)
Comment (xerrni) [reply cmd/mount/mount.go]: No problem. I will do it.

**Comment by xerrni** [comment]
@ncw 
Can you revisit changes? 
I revisited my commit using your tips.

File: cmd/mountlib/mount.go [hunk 3]
```diff
@@ -22,6 +25,30 @@ var (
...
+
+	_, fpErr = fp.Readdirnames(1)
+
+	// directory is not empty
+	if fpErr != io.EOF {
+		var e error
+		if fpErr == nil {
+			e = errors.New("Directory is not empty: " + mountpoint)
+		} else {
+			e = errors.Wrap(fpErr, "Directory is not empty: " + mountpoint)
```
Comment (xerrni) [inline cmd/mountlib/mount.go]: This line has styling error:
`-			e = errors.Wrap(fpErr, "Directory is not empty: " + mountpoint)`
`+			e = errors.Wrap(fpErr, "Directory is not empty: "+mountpoint)`

but I think it is false positive error.

File: cmd/mountlib/mount.go [hunk 4]
```diff
@@ -22,6 +25,30 @@ var (
 	ExtraFlags         *[]string
 )
 
+// Check is folder is empty
+func checkMountEmpty(mountpoint string) error {
+	fp, fpErr := os.Open(mountpoint)
+	defer fs.CheckClose(fp, &fpErr)
+
+	if fpErr != nil {
+		return errors.Wrap(fpErr, "Can not open: " + mountpoint)
```
Comment (xerrni) [inline cmd/mountlib/mount.go]: This line has styling error:
`-		return errors.Wrap(fpErr, "Can not open: " + mountpoint)`
`+		return errors.Wrap(fpErr, "Can not open: "+mountpoint)`

but I think it is false positive error.

File: cmd/mountlib/mount.go (hunk 3 above)
Comment (ncw) [reply cmd/mountlib/mount.go]: Just run the code through [gofmt](https://golang.org/cmd/gofmt/) - that is the go way!  Or even better get [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports) to run on save in your editor.

Comment (xerrni) [reply cmd/mountlib/mount.go]: @ncw 
So should I correct it or leave it as is.
Personally I think code looks better as is right now. ( I`m new to go and I`m basing my opinion on coding style roles in other languages).

Comment (ncw) [reply cmd/mountlib/mount.go]: Correct it please, all code in rclone has been run through gofmt - you'll find the same in nearly all go projects.  There is only one correct way to format go programs - whatever gofmt outputs!

File: cmd/mountlib/mount.go [hunk 5]
```diff
@@ -22,6 +25,30 @@ var (
 	ExtraFlags         *[]string
 )
 
+// Check is folder is empty
+func checkMountEmpty(mountpoint string) error {
+	fp, fpErr := os.Open(mountpoint)
+	defer fs.CheckClose(fp, &fpErr)
```
Comment (ncw) [inline cmd/mountlib/mount.go]: That defer fs.CheckClose needs to go after the if fpErr != nil check I just noticed.  fp will be nil if fpErr != nil.

**Comment by xerrni** [comment]
@ncw 
Can you make review ?

**Comment by ncw** [comment]
Very nice - thank you :-)

**Pull request merged**